c++/src/kj/memory.h

```diff
@@ -656,6 +697,172 @@ class SpaceFor {
   };
 };
 
+// =======================================================================================
+// Pin<T>
+
+template <typename T>
+class Ptr;
+
+template <typename T>
+class Pin {
+// Pin<T> is a smart, in-place storage for T. 
+//
+// Pin<T> should be created on the stack or used as a data member. It should not be 
+// allocated on the heap.
+// Pin<T> is integrated with Ptr<T>, and is legal to move/destroy only when there are no active
+// pointers. 
+// When KJ_ASSERT_PTR_COUNTERS is defined, pointers are tracked and validity of these 
+// operations are asserted.
+// Zero-overhead replacement for T if KJ_ASSERT_PTR_COUNTERS is not defined.
+
+public:
+  template <typename... Params>
+  inline Pin(Params&&... params) : t(kj::fwd<Params>(params)...) {  }
+  // Create new Pin<T> using corresponding T constructor.
+
+  inline Pin(Pin<T>&& other): t(kj::mv(other.t)) {
+  // Move T's ownership. Illegal when live pointers exist.
+#ifdef KJ_ASSERT_PTR_COUNTERS
+    other.ptrCounter.assertEmpty();
+#endif
+  }
+
+  inline ~Pin() {
+  // Destroy a Pin with underlying object. Illegal when live pointers exist.
```

Comment:
non-blocking nit: for the documentation on this, it's worth stressing that the "Illegal when live pointers exist" is only actually checked when `KJ_ASSERT_PTR_COUNTERS` is defined.

Reply:
technically it is still illegal. I made some changes to wording:

```
  // Undefined behavior when live pointers exist, asserted when KJ_ASSERT_PTR_COUNTERS is defined.
```

is this better?